Tennis-scheduler/src/main/java/com/tennis/tennisscheduler/controllers/TimeslotController.java

```diff
@@ -53,23 +66,47 @@ public ResponseEntity<TimeslotDto> getById(@PathVariable long id){
     @PreAuthorize("hasAnyRole('ADMIN','TENNIS_PLAYER')")
     @PostMapping(value = "/")
     public ResponseEntity<TimeslotResponseDto> save(@RequestBody @Valid TimeslotDto timeslotNew, BindingResult result){
+        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+        Person user = (Person)authentication.getPrincipal();
+
         TimeslotResponseDto timeslotResponse = new TimeslotResponseDto();
-        if (!result.hasErrors()){
+        if (!result.hasErrors() && user.getRole().getRoleName().equals(UserType.ROLE_TENNIS_PLAYER)){
+            timeslotNew.personId = user.getId();
             timeslotResponse = timeslotResponseDtoMapper.toTimeslotResponseDto(timeslotService.reserveTimeslot(timeslotDtoMapper.fromTimeslotDtoToTimeslot(timeslotNew)));
             return new ResponseEntity<>(timeslotResponse, HttpStatus.CREATED);
         }
-        else {
+        else if(!result.hasErrors() && user.getRole().getRoleName().equals(UserType.ROLE_ADMIN)){
+            timeslotResponse = timeslotResponseDtoMapper.toTimeslotResponseDto(timeslotService.reserveTimeslot(timeslotDtoMapper.fromTimeslotDtoToTimeslot(timeslotNew)));
+            return new ResponseEntity<>(timeslotResponse, HttpStatus.CREATED);
+        } else {
             timeslotResponse.message = result.getAllErrors();
             return new ResponseEntity<>(timeslotResponse, HttpStatus.BAD_REQUEST);
         }
     }
 
     @PreAuthorize("hasAnyRole('ADMIN','TENNIS_PLAYER')")
     @PutMapping(value = "/{id}")
-    public ResponseEntity<TimeslotResponseDto> update(@PathVariable long id, @RequestBody TimeslotDto timeslotUpdate){
+    public ResponseEntity<TimeslotResponseDto> update(@PathVariable long id, @RequestBody @Valid TimeslotDto timeslotUpdate){
+        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+        Person user = (Person)authentication.getPrincipal();
+
         Timeslot timeslotExisting = timeslotService.getById(id);
-        if (timeslotExisting == null) {
+        if (timeslotExisting == null)
             return new ResponseEntity<>( HttpStatus.NOT_FOUND);
+
+        if(user.getRole().getRoleName().equals(UserType.ROLE_TENNIS_PLAYER)){
```

Comment:
This needs to be refactored. Think about the cases. Here you are returning 401 twice, can that be written in a better way?